Support Bech32 keys in CLI (and rework the implementation to be less fancy)

Anviking opened the pull request:
# Issue Number

#1603 

Alternative implementation to PR #1591 

# Overview
 
- [x] I removed the inspect, child, public fields from `CliKeyScheme` which was the cause for several abstractions.
- [x] I replaced the roundtrip property with more basic ones.
- [x] This opens up for more untested mistakes in the CLI handlers themselves. But I tried to counteract this using point-free notation to avoid potential mixup of `k`, `k'`, `key``, etc.
- [x] Added integration tests making sure `key child`, `key public`, `key inspect` accept keys from stdin. Integration tests also hit both hex and bech32 encodings, as well as public and private derivation.

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Radically re-work CliKeyScheme & add bech32 encoding

This commit abandons the philosophy of having all logic in CliKeyScheme
and all encoding happening using mapKey.

Also: Support reading key inputs from stdin

Only con is that ommitting the key now no longer displays help-text
which could be confusing for users.

== Fixups ==

=== use T.encodeUtf8 / T.decodeUtf8

...rather than e.g. `B8.pack . T.unpack`.

Was told it was the proper way.

=== rename non-recursive `go` to `action`

If we were to only use stdin for key inputs, we could probably use
do-notation, and not have to use this "yet-another" helper.

=== parse simple `KeyEncoding` instead of function

Anviking commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -866,14 +765,13 @@ cmdKeyPublic =
-        $ defaultScheme Icarus
-
+    exec (KeyPublicArgs key) =
+        orReadStdin key
```
This `orReadStdin key` feels a bit strange.

`key` is the key passed as argument. `orReadStdin key` reads the key from stdIn if the arguments are empty.

It's easy to forget it, or make the mistake of using `key` directly.

Maybe it can be moved to `cmd`.

KtorZ replied:
Or, simply make it the only way to provide a value. So, for key commands, inputs HAVE to be provided via stdin, and it's very easy to redirect "literals" if needed to make it like standard arguments:

```
echo ... | cardano-wallet key child
```

Anviking commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -864,36 +863,33 @@ spec = do
-        it "scheme == mapKey (fromBS . toBS) scheme" $ property
-            $ propCliKeySchemeCodecRoundTrip keyByteStringCodec
+        describe "Key Encoding Roundtrips" $ do
+            let decodeAny x = detectEncoding x >>= flip decodeKey x
```
Should use the existing `decodeAnyKey `

Anviking committed:
Factor out test helpers to have HasCallStack

Anviking committed:
Add integration test for piping of CLI key commands

I compared with jcli.

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -367,114 +377,139 @@ runCli = join . customExecParser preferences
+        , "bech32- and hexadecimal encodings are supported."
+        , "\n\n"
+        , "For instance:\n"
+        , "$ cardano-wallet-byron key root --wallet-style icarus --encoding bech32 -- express theme celery coral permit ... \\\n"
```
We should avoid hard-coding the name of the binary here, since it actually depends on where this command is mounted. I'd suggest to simply stick with the commands:

```
key root --wallet-style icarus ...
```

rvl replied:
If you really want the program name, I think it would be fine to do `unsafePerformIO getProgName`.

Anviking replied:
👍

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -367,114 +377,139 @@ runCli = join . customExecParser preferences
+detectEncoding :: Text -> Either String KeyEncoding
+detectEncoding x
+    | "xprv1" `T.isPrefixOf` x = pure Bech32
+    | "xpub1" `T.isPrefixOf` x = pure Bech32
```
Detecting bech32 encoding by detecting the prefix is sort of wrong. A value could be bech32 encoded but simply has a wrong prefix. Yet, a value can't be both bech32 and hex encoded, so we may as well simply try to decode them, and then, check for the prefix in case it's bech32.

Anviking replied:
The problem is that I want `valid_xprv1 <> "ö"` to fail with a `bech32` specific error message, but `0000` to fail with a hex-specific one.

Not sure how to achieve that otherwise.

I experimented some with:

```haskell
    let r1 = (,Hex) <$> decodeHex txt
    let r2 = (,Bech32) <$> decodeBech32 txt
    case (r1, r2) of
        (Right x, _) -> return x
        (_, Right x) -> return x
        (Left (Recognized e), Left (UnrecognizedEncoding _))
            -> Left e
        (Left (UnrecognizedEncoding _), Left (Recognized e))
            -> Left e
        (Left (UnrecognizedEncoding _), Left (UnrecognizedEncoding _))
            -> Left fullKeyEncodingDescription
        (Left (Recognized _), Left (Recognized _))
            -> Left "decodeAnyKey: internal error. The key is recognized as \
                    \bech32- and hex-encoded at the same time, but fails with \
                    \both."
```
but it didn't seem to work out fully.

Anviking replied:
I added golden tests for failures in 1d0cc5effcf02321e0ef0f3d3069d54328d5c726

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -367,114 +377,139 @@ runCli = join . customExecParser preferences
+        dpErr = "Internal error: Unable to extract bytes from bech32 data part"
+
+decodeHex :: Text -> Either String ByteString
+decodeHex txt = fromHex $ B8.pack $ T.unpack . T.strip $ txt
```
This works because hex-encoded characters are ASCII characters, but the "good way" of doing this would be by using `encodeUtf8` from `Data.Text.Encoding`.

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -367,114 +377,139 @@ runCli = join . customExecParser preferences
+
+encodeKey :: KeyEncoding -> XPrvOrXPub -> Either String Text
+encodeKey enc key = case enc of
+    Hex -> T.pack . B8.unpack . hex <$> bytes
```
Same as above, see `Data.Text.Encoding`.

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -612,65 +638,71 @@ newCliKeyScheme = \case
+    return . AXPrv
+        $ CC.deriveXPrv
+            CC.DerivationScheme2
+            (mempty :: Passphrase "Encryption")
```
Hmmm. More like `mempty :: ScrubbedBytes` or `mempty :: ByteString` but, why not. Since `Passphrase` has a `ByteArrayAccess`, it probably doesn't matter.

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -755,14 +774,13 @@ cmdKeyInspect =
+            >>= TIO.putStrLn
+      where
+        go :: Text -> Either String Text
+        go = fmap (\(k, _) -> inspect k) . decodeAnyKey
```
small note but `go` is typically used for recursive functions. So it feels like something is going to recurse but it isn't. Perhaps `action` or `run` is a better fit.

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -1732,8 +1750,23 @@ walletStyleOption defaultStyle accepted = option (eitherReader fromTextS)
     prettyStyle s =
         "  " ++ T.unpack (toText s) ++ " (" ++ fmtAllowedWords s ++ ")"
 
+instance FromText (XPrvOrXPub -> Either String Text) where
```
This is, interesting. Why not simply have an instance from `Text` to `Encoding` ? The encoding as a data-type is perfectly fine to carry on especially because the `encodeKey` arguments are homogeneous here. So this looks like needlessly complex.

Anviking replied:
👍 Think I didn't have `KeyEncoding` in the same sense when I added this.

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -612,65 +638,71 @@ newCliKeyScheme = \case
+
+toPublic :: XPrvOrXPub -> Either String XPrvOrXPub
+toPublic (AXPrv xprv) = return . AXPub . CC.toXPub $ xprv
+toPublic (AXPub _) = Left "Input is already a public key."
```
Wouldn't it be find though to just have an identity here :thinking: ? It's kinda nicer for building scripts that accepts either a private or public key. There's nothing wrong in practice although, it's nice to also have an explicit message indicating the user he's probably doing something he didn't intended. 

No strong opinion here, but just a thought.

Anviking replied:
It occurred to me, but went with this because of:
>  it's nice to also have an explicit message indicating the user he's probably doing something he didn't intended

But I can't really imagine any realistic scenarios motivating either decision.

Me trying really hard to come up with a real argument for erroring:
```
# someone makes a CLI program which accepts both private and public keys
publishAddressOnline = mkAddress | cardano-wallet key public

# The authors forget that the program also accepts private keys

# The authors remove the public command
publishAddressOnline = mkAddress

# Users still think the program accepts private keys

# mkAddress is somehow flawed and leaks the users' private key
```

Anviking replied:
Maybe let's wait and see.

If we realise it will be useful for say, the address CLI commands, then let's add it then.

KtorZ reviewed (approved):
Looks good, although I have some reserve about the dual argument / stdin support. I think we should only have one way for passing inputs (stdin) and stick to it.

rvl commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -367,114 +377,139 @@ runCli = join . customExecParser preferences
-    }
+eitherToIO :: Either String a -> IO a
+eitherToIO (Right a) = return a
+eitherToIO (Left e) = hPutStrLn stderr e >> exitFailure
```
```suggestion
eitherToIO = either die return
```

rvl reviewed (commented):
Looks good

Anviking committed:
Fixup: ensure CLI example uses the correct progName

Anviking committed:
Fixup: simplify using `die` in `eitherToIO`

Co-Authored-By: Rodney Lorrimar <rodney.lorrimar@iohk.io>

Anviking committed:
Add golden tests for key decoding error messages

Anviking committed:
Split keyFromBytes into xprv and xpub specific function

When decoding hex keys it doesn't matter.

But when decoding bech32, we want to make sure that the "xprv" hrp leads
to the bytes being decoded as a xprv and "xpub" as a xpub.

A public key with the "xprv" hrp should not be accepted, or vice versa.
